packages/workspace/src/browser/workspace-service.ts

```diff
@@ -271,27 +297,41 @@ export class WorkspaceService implements FrontendApplicationContribution {
         if (!this.opened) {
             throw new Error('Folder cannot be removed as there is no active folder in the current workspace.');
         }
-        const config = await this.getWorkspaceConfig();
-        if (config) {
-            await this.writeRootFolderConfigFile(
-                config.stat, this._roots.filter(root => uris.findIndex(u => u.toString() === root.uri) < 0)
+        if (this._workspace) {
+            this._workspace = await this.writeWorkspaceFile(
+                this._workspace, this._allRoots.filter(root => uris.findIndex(u => u.toString() === root.uri) < 0)
+            );
+        }
+    }
+
+    private async writeWorkspaceFile(workspaceFile: FileStat | undefined, rootFolders: FileStat[]): Promise<FileStat | undefined> {
+        if (workspaceFile) {
+            const workspaceData = WorkspaceData.transformToRelative(
+                WorkspaceData.buildWorkspaceData(rootFolders.map(f => f.uri)), workspaceFile
             );
+            if (workspaceData) {
+                const stat = await this.fileSystem.setContent(workspaceFile, JSON.stringify(workspaceData));
+                return stat;
+            }
         }
     }
 
-    private async writeRootFolderConfigFile(rootConfigFile: FileStat, rootFolders: FileStat[]): Promise<void> {
-        const folders = rootFolders.slice();
-        if (folders.length === 0 && this._workspaceFolder) {
-            folders.push(this._workspaceFolder);
+    private async getTemporaryWorkspaceFile(): Promise<FileStat | undefined> {
```

Comment:
How about have in memory workspace instead of creating files in the user directories?

Reply:
Sven talked about it a few weeks ago :)

Yes we can keep it in the memory, but that introduces the following issue:

Workspace data stays in the memory and will be missing when theia reloads. One way to stop the browser closing is to override `window.onbeforeunload()`. However, the pop-up dialog from `window.onbeforeunload()` is browser-specific and we don't have much room to customise.

It is fine that the user really doesn't want to save workspace data before reloading theia. But the browser can possibly be closed due to many other reasons. Sometimes it is even beyond the user's control (especially on Windows). Saving the workspace data in a temporary file helps users to restore the status of their workspaces.

Reply:
We send the last recently used workspace before unload, we should send the data instead of a pointer to a file when multi-root workspace has not been saved yet.

Reply:
@svenefftinge 
> we should send the data instead of a pointer (URI) to a file

please correct me if i am wrong - if we send the unsaved data to a file, then i don't see the difference between the 2 solutions - the unsaved data got to be saved somewhere in the filesystem anyway. 

Also, like i mentioned, the `WorkspaceService.stop()` and `WorkspaceService.close()` are not necessarily called, if the browser is closed due to reasons like browser crash or system loses power.

Reply:
@elaihau and me discussed and played a bit with a sharing scenario. And persitening it in a file eagerly, helps because other users immediately see changes made to the list of workspace roots. So I'd like to keep the approach proposed here.